x/authz/keeper/msg_server.go

```diff
@@ -0,0 +1,94 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/authz/types"
+)
+
+var _ types.MsgServer = Keeper{}
+
+// GrantAuthorization implements the MsgServer.GrantAuthorization method.
+func (k Keeper) GrantAuthorization(goCtx context.Context, msg *types.MsgGrantAuthorization) (*types.MsgGrantAuthorizationResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	grantee, err := sdk.AccAddressFromBech32(msg.Grantee)
+	if err != nil {
+		return nil, err
+	}
+	granter, err := sdk.AccAddressFromBech32(msg.Granter)
+	if err != nil {
+		return nil, err
+	}
+
+	authorization := msg.GetGrantAuthorization()
+	// If the granted service Msg doesn't exist, we throw an error.
+	if k.router.Handler(authorization.MethodName()) == nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "%s doesn't exist.", authorization.MethodName())
+	}
+
+	err = k.Grant(ctx, grantee, granter, authorization, msg.Expiration)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventGrantAuthorization,
+			sdk.NewAttribute(types.AttributeKeyGrantType, authorization.MethodName()),
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(types.AttributeKeyGranterAddress, msg.Granter),
+			sdk.NewAttribute(types.AttributeKeyGranteeAddress, msg.Grantee),
+		),
+	)
+
+	return &types.MsgGrantAuthorizationResponse{}, nil
+}
+
+// RevokeAuthorization implements the MsgServer.RevokeAuthorization method.
+func (k Keeper) RevokeAuthorization(goCtx context.Context, msg *types.MsgRevokeAuthorization) (*types.MsgRevokeAuthorizationResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	grantee, err := sdk.AccAddressFromBech32(msg.Grantee)
+	if err != nil {
+		return nil, err
+	}
+	granter, err := sdk.AccAddressFromBech32(msg.Granter)
+	if err != nil {
+		return nil, err
+	}
+
+	err = k.Revoke(ctx, grantee, granter, msg.MethodName)
+	if err != nil {
+		return nil, err
+	}
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventRevokeAuthorization,
```

Comment:
Shouldn't these events be emitted in the keeper?

One example is when the SendAuthorization limit is reached, then the authorization is automatically revoked (see Accept from SendAuthorization). I would expect this scenario to emit an event. However, by putting event emission in the msg_server, there's no event emitted in this scenario.